### ISSUE TITLE
Move all metrics from core/cv/metrics to core/metrics?

### ISSUE THREAD

**Issue by RishabhGarg108** [mlpack/mlpack#2981]
Currently, we have most of the common metrics used to evaluate machine learning models under `core/cv/metrics`. The issue is that the metrics implemented under `cv/metrics` have only a single `Evaluate` method that takes a trained model, test data and test labels. First, it makes predictions on the given model and then calculates the metric. The current implementation is done only by keeping in mind the requirement for cross-validation.

But, we can also provide another method that takes predictions and true labels to evaluate the metric. 
Why?---because this is the actual job of a metric. This can be very useful both inside and outside mlpack. Outside, users can use it when they have predictions and labels. While inside, it can be used in the unit tests.

Also, after making this change it would make sense to shift all these to `core/metrics`.

**Comment by rcurtin** [comment]
The thing is, the stuff that's currently in `metrics/` are actually metric functions: https://en.wikipedia.org/wiki/Metric_(mathematics)

However, I suppose we could rename those to "distance functions", which might be more familiar to users not from the nearest neighbor / metric spaces communities.

Any thoughts on that?

**Comment by RishabhGarg108** [comment]
Hey @rcurtin, "distance functions" can be used, but I would say "metrics" is a better name for two reasons:-

1. According to the Wikipedia definition you provided, a metric is a function that determines the distance between two quantities. In our use cases, all we care about is the distance between the predictions and actual labels. So, this definition applies perfectly to the stuff inside `cv/metrics/`.
2. All the other libraries like sklearn, tensorflow and pytorch have all these under metrics submodule. So, it will be easy to learn for a person coming from that background.

**Comment by rcurtin** [comment]
Well, so what I mean is that what's currently in `metrics/` is fundamentally different than what's in `cv/metrics/`.  What's in `metrics/` now are metrics according to the Wikipedia page I linked to---specifically, they satisfy the triangle inequality, and can be used for tasks like nearest neighbor search and others.

But, for computing performance on a model (which is what `cv/metrics/` does), that does not satisfy the triangle inequality, and is a metric only in the "colloquial" definition, not in the mathematical definition on the Wikipedia page.

So, my proposal is just to move what's currently in `metrics/` to `distances/`, and then `cv/metrics/` can move to `metrics/`.  This may be more familiar to users who are used to other toolkits, like you pointed out, and `distances` is a name that's also used for metrics that satisfy the triangle inequality, so I think there will be no confusion there either. :+1:

**Comment by RishabhGarg108** [comment]
Ahh, I see. It does make sense to me now. Thanks for the clarification :). Now, I too agree that we can name it `distances/` 👍
